azerothcore/azerothcore-wotlk
# fix(Core/Creature) Make `Respawn()` respect conditions and linked respawns

## Changes Proposed:

This PR proposes changes to:

This PR changes the `Creature::Respawn()` function to respect the following features:
- [`linked_respawn`](https://www.azerothcore.org/wiki/linked_respawn) entries
- [`conditions`](https://www.azerothcore.org/wiki/conditions) entries

It accomplishes this by moving the logic to check all of these from `Creature::Update()` to `Creature::Respawn()`. That means that any outside class/function that calls `Respawn()` will now benefit from these checks that were previously not applied to them.

**Requirements for testing**
- requires #17590 (to make Alexei's adds have `creature_formation` and `linked_respawn` entries)

Previously, the creature was forced to respawn regardless of any conditions or linked respawns. This change reworks the `Respawn()` function to check for conditions or linked respawns that should be respected before the respawn occurs, unless the respawn is forced (like in the GM `.revive` command).

## Issues Addressed:

- Related to https://github.com/azerothcore/azerothcore-wotlk/issues/8239
- Will resolve issues with https://github.com/azerothcore/azerothcore-wotlk/pull/17590

## SOURCE:

The changes have been validated through:

## Tests Performed:

This PR has been:


## How to Test the Changes:



**Requirements**
- requires #17590 (to make Alexei's adds have `creature_formation` and `linked_respawn` entries)

_Test add respawns on reset_
1. `.go xyz 168.953613 -57.557426 75.397003 289 5.036365`
2. Pull the boss and kill one add
3. `.gm on`
4. Observe the dead add being respawned correctly

_Test adds not respawning on boss dead_
1. `.go xyz 168.953613 -57.557426 75.397003 289 5.036365`
2. Pull the boss and kill one add, then kill the boss.
3. `.gm on`
4. Observe the remaining add running back home, and the other add not respawning.

## Known Issues and TODO List:



## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
